modules/src/ics02_client/handler/update_client.rs

```diff
@@ -208,4 +246,362 @@ mod tests {
+            AnyHeader::Tendermint(mut theader) => {
+                let cons_state = ctx
+                    .latest_consensus_states(&client_id, &client_height)
+                    .clone();
+                if let AnyConsensusState::Tendermint(tcs) = cons_state {
+                    theader.signed_header.header.time = tcs.timestamp;
+                }
+                AnyHeader::Tendermint(theader)
+            }
+            AnyHeader::Mock(header) => AnyHeader::Mock(header),
+        };
+
+        let msg = MsgUpdateAnyClient {
+            client_id: client_id.clone(),
+            header: latest_header,
+            signer,
+        };
+
+        let output = dispatch(&ctx, ClientMsg::UpdateClient(msg.clone()));
+
+        match output {
+            Ok(HandlerOutput {
+                result,
+                mut events,
+                log,
+            }) => {
+                assert_eq!(events.len(), 1);
+                let event = events.pop().unwrap();
+                assert!(
+                    matches!(event, IbcEvent::UpdateClient(e) if e.client_id() == &msg.client_id)
+                );
+                assert!(log.is_empty());
+                // Check the result
+                match result {
+                    Update(upd_res) => {
+                        assert_eq!(upd_res.client_id, client_id);
+                        assert!(!upd_res.client_state.is_frozen());
+                        assert_eq!(
+                            upd_res.client_state,
+                            ctx.latest_client_states(&client_id).clone()
+                        );
+                        assert_eq!(upd_res.client_state.latest_height(), msg.header.height(),)
+                    }
+                    _ => panic!("update handler result has incorrect type"),
+                }
+            }
+            Err(_err) => {
+                panic!("unexpected error");
+            }
+        }
+    }
+
+    #[test]
+    fn test_update_synthetic_tendermint_client_duplicate_height_frozen() {
+        let client_id = ClientId::new(ClientType::Tendermint, 0).unwrap();
+        let client_height = Height::new(1, 20);
+
+        let chain_start_height = Height::new(1, 11);
+
+        let ctx = MockContext::new(
+            ChainId::new("mockgaiaA".to_string(), 1),
+            HostType::Mock,
+            5,
+            chain_start_height,
+        )
+        .with_client_parametrized(
+            &client_id,
+            client_height,
+            Some(ClientType::Tendermint), // The target host chain (B) is synthetic TM.
+            Some(client_height),
+        );
+
+        let signer = get_dummy_account_id();
+
+        let block_ref = HostBlock::generate_block(
+            ChainId::new("mockgaiaB".to_string(), 1),
+            HostType::SyntheticTendermint,
+            client_height.revision_height,
+        );
+
+        let latest_header: AnyHeader = Some(block_ref).map(Into::into).unwrap();
```

Comment:
On my side this particular test fails with:

```
test ics02_client::handler::update_client::tests::test_update_synthetic_tendermint_client_duplicate_height_frozen ... FAILED

failures:

---- ics02_client::handler::update_client::tests::test_update_synthetic_tendermint_client_duplicate_height_frozen stdout ----
thread 'ics02_client::handler::update_client::tests::test_update_synthetic_tendermint_client_duplicate_height_frozen' panicked at 'unexpected error: header verification failed with reason: consensus state not found at: 07-tendermint-0 at height 0-0
```

I have a suspicion that this conversion either loses data or doesn't set the trusted height appropriately, but I'll have to take another look at it in the morning. Perhaps @cezarad can give some context on this?

Reply:
We can erase this test. It was checking that the client is frozen when an update for an already updated height arrives, which is deprecated.